sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -731,8 +717,13 @@ public void copy(List<String> srcFilenames, List<String> destFilenames) throws I
     }
 
     @Override
-    public void remove(Collection<String> filenames) throws IOException {
-      gcsUtil.remove(filenames);
+    public void removeDirectoryAndFiles(String directory) throws IOException {
+      IOChannelFactory factory = IOChannelUtils.getFactory(directory);
+      Collection<String> matches = factory.match(directory + "/*");
+      LOG.debug("Removing {} temporary files found under {}", matches.size(), directory);
+      gcsUtil.remove(matches);
+      // No need to remove the directory itself: GCS doesn't have directories, so if the directory
+      // is empty, then it already doesn't exist.
```

Comment:
Fair. Added that. Still not perfect, but at least it will be perfect in pipelines without bundle failures, which is, hopefully, most pipelines.